src/models/metrics.py

```diff
@@ -0,0 +1,137 @@
+"""Functions for calculating metrics from a GeoGraph."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Optional
+
+import networkx as nx
+import numpy as np
+
+if TYPE_CHECKING:
+    from src.models import geograph
+
+
+# define a metric dataclass with < <= => > == comparisons that work as you would
+# expect intuitively
+@dataclass()
+class Metric:
+    """Class to represent a metric for a GeoGraph, with associated metadata."""
+
+    value: Any  # No good Numpy type hints
+    name: str
+    description: str
+    variant: Optional[str]
+    unit: Optional[str] = None
+
+    def __eq__(self, o: object) -> bool:
+        if not isinstance(o, Metric):
+            return False
+        return self.value == o.value
+
+    def __lt__(self, o: object) -> bool:
+        if not isinstance(o, Metric):
+            return False
+        return self.value < o.value
+
+    def __le__(self, o: object) -> bool:
+        if not isinstance(o, Metric):
+            return False
+        return self.value <= o.value
+
+    def __gt__(self, o: object) -> bool:
+        if not isinstance(o, Metric):
+            return False
+        return self.value > o.value
+
+    def __ge__(self, o: object) -> bool:
+        if not isinstance(o, Metric):
+            return False
+        return self.value >= o.value
+
+
+def _num_components(geo_graph: geograph.GeoGraph) -> Metric:
+    return Metric(
+        value=nx.number_connected_components(geo_graph.graph),
+        name="num_components",
+        description="The number of connected components in the graph.",
+        variant="component",
+    )
+
+
+def _avg_patch_area(geo_graph: geograph.GeoGraph) -> Metric:
+    return Metric(
+        value=np.mean(geo_graph.df.area.values),
+        name="avg_patch_area",
+        description="The average area of the patches in the graph.",
+        variant="conventional",
+    )
+
+
+def _total_area(geo_graph: geograph.GeoGraph) -> Metric:
+    return Metric(
+        value=np.sum(geo_graph.df.area.values),
+        name="total_area",
+        description="The total area of all the patches in the graph.",
+        variant="conventional",
+    )
+
+
+def _avg_component_area(geo_graph: geograph.GeoGraph) -> Metric:
+    comp_geograph = geo_graph.components
+    if not comp_geograph.has_df:
+        raise ValueError(
+            "This metric is not valid for ComponentGeoGraphs without a dataframe."
+        )
+    return Metric(
+        value=np.mean(comp_geograph.df.area.values),
+        name="avg_component_area",
+        description="The average area of the components in the graph",
+        variant="component",
+    )
+
+
+def _avg_component_isolation(geo_graph: geograph.GeoGraph) -> Metric:
+    """Calculate the average distance to the next-nearest component."""
+    comp_geograph = geo_graph.components
```

Comment:
Great that we can compute this, at least for smaller habitats.

Again usage-question: how to I make sure that a `GeoGraph` instance has all the relevant attributes to be able to compute this?

Reply:
Also: how hard would it be to extend this to just regular patch isolation as well?

Reply:
You can do `geograph.components = geograph.get_graph_components(calc_polygons=True, add_distance_edges=True)`, to calculate the distance edges and make this metric work, but as warned this will be very expensive for graphs with more than about 100 components.

Reply:
For patch isolation I think the approach I've been taking to calculate distances between everything is too inefficient, so I think you would have to change the idea of the metric from being distance to next-nearest patch to being average distance between neighbouring patches.

Reply:
Also also: should we set this to zero if there is only a single component? I think this would make sense, as we would want this metric to go down as more components become connected, and in the limit (i.e. all components are connected) it should be the lowest = 0.

Reply:
Good idea, done in f5b80c0

Reply:
Thanks for the info, I think the best quick addition for this PR will be to add this fix 

```(geograph.components = geograph.get_graph_components(calc_polygons=True, add_distance_edges=True)```

 to the warning and docstring of the two component metric functions. Then as a next step we could extend the get components function as I suggested above.

Reply:
(Again beyond the scope of this PR) About patch and component isolation: didn't give this extensive thought but I think we can use the existing graph to make this more information:
1. only computing distances where there are edges (because if there is an edge, the minimal distance patch must has such an edge)
2. slowly searching for nearby patches (like creating a larger max_travel_distance habitat) by adding a double_max travel distance buffer and then adding the edges with distance to the graph, and doing this in incriments of the max_travel_distance, until we found an edge.
Together this should still be more efficient in most cases than the complete distance computation at the moment.